### PR TITLE
X11 forwarding on Mac + docker 

### PR DESCRIPTION
**Pull request by kkauder**
### Briefly, what does this PR introduce?
Forward X11 from inside the docker container on macOS
Use cases: Use container root, useful where web-based display isn't enough; other GUIs, such as GEANT4 (not tested, may need OpenGL fiddling)
- Note: not needed with singularity where X11 forwarding works out of the box

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ x ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Documentation has been added / updated --> See below for Notes
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No. None. 

### Does this PR change default behavior?
No, but see notes

### Notes:
- To avoid any default behavior change, users need to use "-X". This is documented in "-h" which should be added to the README. But the changes are light. To work, users need to check a box one time in the settings. However, the worst case scenario is that X11 doesn't work, not a breaking error.

### CODE REVIEW

**Comment by wdconinc** [comment]
My thoughts: this should be default. But I won't review since I can't test.

**Comment by kkauder** [comment]
@ajentsch Could you give this a quick test? All I did was `new TCanvas`, but maybe you can quickly pull up a Geant GUI or so.

**Comment by ajentsch** [comment]
GEANT4 openGL visualization using npsim works (--runType vis): 
image

TBrowsers also work:

image

**Review by ajentsch — APPROVED** [review]
Changes do what they are advertised to do. It's very nice to be able to open a TBrowser or a simple GEANT event display.

**Comment by kkauder** [comment]
Thanks, Alex!
@wdconinc , only question is, right now it  prints
```
% ./eic-shell -X
Activating XQuartz support in Docker.
If needed (should be only once): In XQuartz settings --> Security --> enable "Allow connections from network clients"
localhost being added to access control list
```
- Line 1 is optional, should probably go away if this becomes the default
- Line 3 should be kept. Can be suppressed by  `>/dev/null/` but informing the user that `xhost +localhost happened` can't hurt.
- Line 2 is the questionable one. It could go into the documentation instead, if people read that. I can also in principle detect from the command line if that option is checked or not. But that's a non-trivial parse of a multiline text that may not be stable through version updates.
Thoughts?

**Comment by wdconinc** [comment]
If it is default, then it should be quiet and print warnings when it fails.

Other points:
- the docker install is only done for Darwin, so no need to check again in the docker function, which probably allows you get rid of some of the EOF2, EOF3, EOF4 fences.

**Comment by kkauder** [comment]
- I'll try to catch problems (not easy off the top of my hand, but I can dig into it for a bit tomorrow or so)
- I see in `install.sh`
```
case ${OS} in
  Linux)
[...]
    if [ "$CPU" = "arm64" ]; then
      install_docker
```
So I prepared for future Linxu ARM boxes.

**Comment by wdconinc** [comment]
> So I prepared for future Linux ARM boxes (and no singularity support for them)

Singularity works fine on linux arm, https://ubuntu.pkgs.org/18.04/ubuntu-universe-arm64/singularity-container_2.4.2-4_arm64.deb.html. This must be something else, like running Linux on a Mac M1 or so. @sly2j ?

**Comment by kkauder** [comment]
ping @sly2j

**Comment by kkauder** [comment]
@wouter, @sly2j , can we just merge as-is, or should we sort out the arm64/Darwin question?

**Comment by kkauder** [comment]
I made X support the default and reduced verbosity
It also now detects whether XQuartz needs to be reconfigured and only squawks if it does.

I think the arm / darwin question can go on the backburner. This is actually a rather much-requested QOL improvement and we should just merge it IMHO.

**Comment by kkauder** [comment]
ping

File: install.sh [hunk 1]
```diff
@@ -340,6 +340,7 @@ function print_the_help {
   echo "USAGE:  ./eic-shell [OPTIONS] [ -- COMMAND ]"
   echo "OPTIONAL ARGUMENTS:"
   echo "          -u,--upgrade    Upgrade the container to the latest version"
+  echo "          --noX           Disable X11 forwarding on macOS (not needed for singularity)"
```
Comment (wdconinc) [inline install.sh:343]: Remove unnecessary parenthetical (you're in `install_docker`).
```suggestion
  echo "          --noX           Disable X11 forwarding on macOS"
```
It would be nice if you could only print this on macOS, or make this work on linux too by not just greping for Xquartz.

File: install.sh [hunk 2]
```diff
@@ -353,15 +354,25 @@ function print_the_help {
 }
 
 UPGRADE=
-
+MACX=1
```
Comment (wdconinc) [inline install.sh:357]: Use consistent approach to flag variables: empty if unset; not empty when set; variable name is option name in uppercase.
```suggestion
NOX=1
```

File: install.sh [hunk 3]
```diff
@@ -353,15 +354,25 @@ function print_the_help {
...
+EOF
+  if [ `uname -s` = 'Darwin' ]; then
+      cat << EOF2 >> eic-shell
+    --noX)
+      MACX=
+      shift
+      ;;
+EOF2
+  fi
+cat << EOF3 >> eic-shell
```
Comment (wdconinc) [inline install.sh:374]: Don't escape the heredoc.

File: install.sh [hunk 4]
```diff
@@ -383,12 +394,28 @@ if [ ! -z \${UPGRADE} ]; then
...
+EOF3
+
+  if [ `uname -s` = 'Darwin' ]; then
+      echo 'if [ ${MACX} ]; then' >> eic-shell
+      # echo '  echo Activating XQuartz support in Docker.' >> eic-shell
+
+# cat << EOF4 >> eic-shell
+#   echo 'If needed (should be only once): In XQuartz settings --> Security --> enable "Allow connections from network clients"'
+# EOF4
+
```
Comment (wdconinc) [inline install.sh:406]: Remove comments or print useful messages if users have to do something.

File: install.sh [hunk 5]
```diff
@@ -383,12 +394,28 @@ if [ ! -z \${UPGRADE} ]; then
...
+      echo ' nolisten=`defaults find nolisten_tcp | grep nolisten | awk ' "'{print" '$3}'"'" '|cut -b 1 `' >> eic-shell
+      echo ' [[ $nolisten -ne 0 ]] && echo "For X support: In XQuartz settings --> Security --> enable \"Allow connections from network clients\" and restart (should be only once)."' >> eic-shell
+      ## getting the following single and double quotes, escapes and backticks right was a nightmare
```
Comment (wdconinc) [inline install.sh:409]: I don't like this change from heredoc to echo into file. It makes this less maintainable.

**Comment by kkauder** [comment]
Ping...

**Comment by wdconinc** [comment]
> Ping...

See comments?

**Comment by kkauder** [comment]
Let me try again to see if I can comment

File: install.sh (hunk 1 above)
Comment (kkauder) [reply install.sh:343]: I removed the parenthetical, it's not needed since we're in the docker function anyway.
To your second point, in linux/singularity no such trickery is needed, from my testing X11 works just fine out of the box.

File: install.sh (hunk 2 above)
Comment (kkauder) [reply install.sh:357]: Done

File: install.sh (hunk 3 above)
Comment (kkauder) [reply install.sh:374]: I don't quite understand the comment, but I'll just remove the conditional and assume docker == darwin

File: install.sh (hunk 4 above)
Comment (kkauder) [reply install.sh:406]: The useful message is a couple of lines above this,
```
      echo ' nolisten=`defaults find nolisten_tcp | grep nolisten | awk ' "'{print" '$3}'"'" '|cut -b 1 `' >> eic-shell
      echo ' [[ $nolisten -ne 0 ]] && echo "For X support: In XQuartz settings --> Security --> enable \"Allow connections from network clients\" and restart (should be only once)."' >> eic-shell
```
Figuring that out was non-trivial btw. 
Cleaned up the comments.

File: install.sh (hunk 5 above)
Comment (kkauder) [reply install.sh:409]: Double-edged sword. I tried with a heredoc and couldn't get it to work. And even if I did, it would add yet more escapes and weird stuff making it on the whole less maintainable again. It's mostly the need to run it through awk. 
If you have a cleaner way to get the ":0" out of
```
 2281 ??         0:00.00 /opt/X11/bin/Xquartz :0 -listen tcp -iglx -auth /Users/eickolja/.serverauth.1971
 ```
 maybe we can change back to a heredoc

**Comment by kkauder** [comment]
What's the workflow after approval? Who executes this merge?

**Review by wdconinc — APPROVED** [review]
